File: csource/ecc_drv/ECDSA.c
/*
 * Copyright 2018 The OpenWallet Authors
 * This file is part of the OpenWallet library.
 *
 * The OpenWallet library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The OpenWallet library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#include "ECDSA.h"

uint16_ow ECDSA_genPubkey(ECC_CURVE_PARAM *curveParam, uint8_ow *prikey, ECC_POINT *pubkey)
{
    ECC_POINT *point_g = NULL;
    
    if(!is_prikey_legal(curveParam, prikey))
        return ECC_PRIKEY_ILLEGAL;
    
    point_g = calloc(1, sizeof(ECC_POINT));
    
    memcpy(point_g -> x, curveParam -> x, ECC_LEN);
    memcpy(point_g -> y, curveParam -> y, ECC_LEN);
    
    if(point_mul(curveParam, point_g, prikey, pubkey))
    {
        free(point_g);
        return ECC_PRIKEY_ILLEGAL;
    }
    
    free(point_g);
    return SUCCESS;
}



uint16_ow ECDSA_sign(ECC_CURVE_PARAM *curveParam, uint8_ow *prikey, uint8_ow *message, uint16_ow message_len,uint8_ow *rand, uint8_ow hash_flag, uint8_ow *sig)
{
    uint8_ow *k = NULL, *tmp = NULL;
    ECC_POINT *point = NULL;
    if(!is_prikey_legal(curveParam, prikey))
        return ECC_PRIKEY_ILLEGAL;
    k = calloc(ECC_LEN, sizeof(uint8_ow));
    tmp = calloc(ECC_LEN, sizeof(uint8_ow));
    point = calloc(1, sizeof(ECC_POINT));
    while(1)
    {
        memcpy(point -> x, curveParam -> x, ECC_LEN);
        memcpy(point -> y, curveParam -> y, ECC_LEN);
        
        if(rand==NULL)
        {
            bigrand_get_rand_range(k, curveParam -> n, prikey, ECC_LEN, message, message_len);
        }
        else
        {
            memcpy(k,rand,ECC_LEN);
        }
        point_mul(curveParam, point, k, point);
        bignum_mod(point -> x, curveParam -> n, sig);
        if(is_all_zero(sig, ECC_LEN))
            continue;
        if(!hash_flag)//传入的是消息
        {
            sha256_hash(message, message_len, tmp);
        }
        else//传入的是哈希值
        {
            if(message_len != ECC_LEN)
            {
                return LENGTH_ERROR;
            }
            else
            {
                 memcpy(tmp, message, message_len);
            }
        }
        bignum_mod_mul(prikey, sig, curveParam -> n, sig + ECC_LEN);
        bignum_mod_add(tmp, sig + ECC_LEN, curveParam -> n, tmp);
        bignum_mod_inv(k, curveParam -> n, k);
        bignum_mod_mul(k, tmp, curveParam -> n, sig + ECC_LEN);
        if(is_all_zero(sig + ECC_LEN, ECC_LEN))
            continue;
        else
            break;
    }
    free(k);
    free(tmp);
    free(point);
    return SUCCESS;
}

uint16_ow ECDSA_verify(ECC_CURVE_PARAM *curveParam, ECC_POINT *pubkey, uint8_ow *message, uint16_ow message_len,uint8_ow hash_flag, uint8_ow *sig)
{
    uint8_ow *tmp1 = NULL, *tmp2 = NULL;
    ECC_POINT *point1 = NULL, *point2 = NULL;
    if(!is_pubkey_legal(curveParam, pubkey))
        return ECC_PUBKEY_ILLEGAL;
    
    if(is_all_zero(sig, ECC_LEN) || memcmp(sig, curveParam -> n, ECC_LEN) >= 0 || is_all_zero(sig + ECC_LEN, ECC_LEN) || memcmp(sig + ECC_LEN, curveParam -> n, ECC_LEN) >= 0)
        return FAILURE;
    
    tmp1 = calloc(ECC_LEN, sizeof(uint8_ow));
    tmp2 = calloc(ECC_LEN, sizeof(uint8_ow));
    point1 = calloc(1, sizeof(ECC_POINT));
    point2 = calloc(1, sizeof(ECC_POINT));
    if(!hash_flag) //需要内部计算哈希值
    {
        sha256_hash(message, message_len, tmp1);
    }
    else//外部已经计算哈希值
    {
        if(message_len != ECC_LEN)
        {
            return LENGTH_ERROR;
        }
        memcpy(tmp1,message,message_len);
    }
    
    bignum_mod_inv(sig + ECC_LEN, curveParam -> n, tmp2);
    bignum_mod_mul(tmp1, tmp2, curveParam -> n, tmp1);
    bignum_mod_mul(sig, tmp2, curveParam -> n, tmp2);
    
    memcpy(point2 -> x, curveParam -> x, ECC_LEN);
    memcpy(point2 -> y, curveParam -> y, ECC_LEN);
    
    point_mul(curveParam, point2, tmp1, point1);
    point_mul(curveParam, pubkey, tmp2, point2);
    
    if(point_add(curveParam, point1, point2, point1))
    {
        free(tmp1);
        free(tmp2);
        free(point1);
        free(point2);
        return FAILURE;
    }
    
    bignum_mod(point1 -> x, curveParam -> n, tmp1);
    
    if(memcmp(tmp1, sig, ECC_LEN))
    {
        free(tmp1);
        free(tmp2);
        free(point1);
        free(point2);
        return FAILURE;
    }
    free(tmp1);
    free(tmp2);
    free(point1);
    free(point2);
    return SUCCESS;
}

uint16_ow ECDSA_recover_public(ECC_CURVE_PARAM *curveParam,uint8_ow *sig,uint32_ow sig_len,uint8_ow *msg,uint32_ow msg_len,uint8_ow hash_flag,uint8_ow *pubkey)
{
    int k=0;
    uint16_ow ret=0x5a5a;
    uint8_ow *buf1=NULL,*buf2=NULL,*r_inv=NULL,*hash=NULL,*s=NULL;
    ECC_POINT *point1,*point2,*R,*G;
    if(!curveParam || !sig || !msg)
    {
        return FAILURE;
    }
    buf1=calloc(65,sizeof(uint8_ow));
    buf2=calloc(33,sizeof(uint8_ow));
    r_inv=calloc(32,sizeof(uint8_ow));
    hash=calloc(32,sizeof(uint8_ow));
    s=calloc(32,sizeof(uint8_ow));
    point1=calloc(1,sizeof(ECC_POINT));
    point2=calloc(1,sizeof(ECC_POINT));
    R=calloc(1,sizeof(ECC_POINT));
    G=calloc(1,sizeof(ECC_POINT));
    if(!hash_flag)
    {
        sha256_hash(msg, msg_len, hash);
    }
    else
    {
        if(msg_len !=32)
        {
            return LENGTH_ERROR;
        }
        else
        {
            memcpy(hash,msg,32);
        }
    }
    memcpy(buf2+1,sig,ECC_LEN);
    bignum_mod_inv(sig, curveParam->n,  r_inv);
    //get point G
    memcpy(G->x,curveParam->x,ECC_LEN);
    memcpy(G->y,curveParam->y,ECC_LEN);
    for(k=0;k<2;k++)
    {
        if(k == 0) //设置R->y为偶数
        {
            buf2[0]=0x02;
        }
        if(k == 1) //设置R->y为奇数
        {
            buf2[0]=0x03;
        }
        if(k == 0)
        {
            if(sig[64]==1) //进行curveParam->n - s的操作，需要将s还原
            {
                bignum_sub(curveParam->n, sig+ECC_LEN, ECC_LEN ,s);
            }
        }
        if(k==1)
        {
            if(sig[64] == 0)//进行curveParam->n - s的操作，需要将s还原
            {
                bignum_sub(curveParam->n, sig+ECC_LEN, ECC_LEN ,s);
            }
        }
        point_decompress(curveParam,  buf2,33, buf1);
        //get point R
        memcpy(R->x,buf1+1,ECC_LEN);
        memcpy(R->y,buf1+1+ECC_LEN,ECC_LEN);
        //point1=s*R
        point_mul(curveParam, R, s,  point1);
        //point2=e*G
        point_mul(curveParam, G, hash, point2);
        //计算 [-1]point2
        bignum_sub(curveParam->p, point2->y, ECC_LEN ,  point2->y);
        point_add(curveParam, point1,  point2,  point1);
        point_mul(curveParam, point1, r_inv,  point2);
        ret=ECDSA_verify(curveParam,  point2, hash,  32, 1,  sig);
        if(ret==SUCCESS)
        {
            memcpy(pubkey,point2->x,ECC_LEN);
            memcpy(pubkey+ECC_LEN,point2->y,ECC_LEN);
            break;
        }
    }
    free(buf1);
    free(buf2);
    free(r_inv);
    free(hash);
    free(point1);
    free(point2);
    free(R);
    free(G);
    free(s);
    return ret;
}

File: ecdsa.go
/*
 * Copyright 2020 The openwallet Authors
 * This file is part of the openwallet library.
 *
 * The openwallet library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The openwallet library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */
package owcrypt

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"errors"
	"math/big"
	"sync"
)

type curveParam struct {
	elliptic.CurveParams
}

var (
	ErrPrivateKeyIllegal = errors.New("Invalid private key data!")
	ErrUnknownCurve      = errors.New("Unknown curve type!")
	ErrMessageIllegal    = errors.New("Invalid message data!")
	ErrPublicKeyIllegal  = errors.New("Invalid public key data!")
)

var (
	initonce sync.Once
	three    = new(big.Int).SetUint64(3)
)

func genPublicKey(privateKey []byte, name string) ([]byte, error) {
	if privateKey == nil || len(privateKey) != 32 {
		return nil, ErrPrivateKeyIllegal
	}

	var curve *curveParam
	var k1curve *secp256k1Curve
	var sm2Curve *sm2_stdCurve
	privateKeyBig := new(big.Int).SetBytes(privateKey)
	priv := new(ecdsa.PrivateKey)

	if privateKeyBig.Cmp(big.NewInt(0)) == 0 {
		return nil, ErrPrivateKeyIllegal
	}
	priv.D = privateKeyBig
	if name == "secp256k1" {
		k1curve = secp256k1

		if privateKeyBig.Cmp(k1curve.Params().N) >= 0 {
			return nil, ErrPrivateKeyIllegal
		}

		priv.PublicKey.Curve = k1curve

		priv.PublicKey.X, priv.PublicKey.Y = k1curve.ScalarBaseMult(privateKey)

	} else if name == "sm2_std" {

		sm2Curve = sm2_std

		if privateKeyBig.Cmp(sm2Curve.Params().N) >= 0 {
			return nil, ErrPrivateKeyIllegal
		}

		priv.PublicKey.Curve = sm2Curve

		priv.PublicKey.X, priv.PublicKey.Y = sm2Curve.ScalarBaseMult(privateKey)

	} else { // ecdsa
		if name == "secp256r1" {
			curve = secp256r1
		} else {
			return nil, ErrUnknownCurve
		}

		if privateKeyBig.Cmp(curve.Params().N) >= 0 {
			return nil, ErrPrivateKeyIllegal
		}

		priv.PublicKey.Curve = curve
		priv.D = privateKeyBig
		priv.PublicKey.X, priv.PublicKey.Y = curve.ScalarBaseMult(privateKey)

	}

	x := priv.PublicKey.X.Bytes()

	for len(x) < 32 {
		x = append([]byte{0x00}, x...)
	}

	y := priv.PublicKey.Y.Bytes()

	for len(y) < 32 {
		y = append([]byte{0x00}, y...)
	}
	return append(x, y...), nil
}

func sign(privateKey, ID, hash []byte, name string) ([]byte, byte, error) {

	if privateKey == nil || len(privateKey) != 32 {
		return nil, 0, ErrPrivateKeyIllegal
	}

	if hash == nil || len(hash) != 32 {
		return nil, 0, ErrMessageIllegal
	}

	var curve *curveParam
	var k1curve *secp256k1Curve
	var sm2Curve *sm2_stdCurve
	privateKeyBig := new(big.Int).SetBytes(privateKey)
	priv := new(ecdsa.PrivateKey)

	if privateKeyBig.Cmp(big.NewInt(0)) == 0 {
		return nil, 0, ErrPrivateKeyIllegal
	}
	priv.D = privateKeyBig
	if name == "secp256k1" {
		k1curve = secp256k1

		if privateKeyBig.Cmp(k1curve.Params().N) >= 0 {
			return nil, 0, ErrPrivateKeyIllegal
		}

		priv.PublicKey.Curve = k1curve

		priv.PublicKey.X, priv.PublicKey.Y = k1curve.ScalarBaseMult(privateKey)

	} else if name == "secp256r1" {
		curve = secp256r1
		if privateKeyBig.Cmp(curve.Params().N) >= 0 {
			return nil, 0, ErrPrivateKeyIllegal
		}

		priv.PublicKey.Curve = curve
		priv.D = privateKeyBig
		priv.PublicKey.X, priv.PublicKey.Y = curve.ScalarBaseMult(privateKey)
	} else if name == "sm2_std" {
		sm2Curve = sm2_std

		if privateKeyBig.Cmp(sm2Curve.Params().N) >= 0 {
			return nil, 0, ErrPrivateKeyIllegal
		}

		priv.PublicKey.Curve = sm2Curve
		priv.D = privateKeyBig
		priv.PublicKey.X, priv.PublicKey.Y = sm2Curve.ScalarBaseMult(privateKey)

		signature := make([]byte, 64)

		r, s, v, err := sm2_std_sign(priv, hash, ID)
		if err != nil {
			return nil, 0, err
		}

		rBytes := r.Bytes()
		sBytes := s.Bytes()

		copy(signature[32-len(rBytes):32], rBytes)
		copy(signature[64-len(sBytes):64], sBytes)

		return signature, v, nil
	} else {
		return nil, 0, ErrUnknownCurve
	}

	signature := make([]byte, 64)

	r, s, v, err := signecdsa(rand.Reader, priv, hash)
	if err != nil {
		return nil, 0, err
	}

	rBytes := r.Bytes()
	sBytes := s.Bytes()

	copy(signature[32-len(rBytes):32], rBytes)
	copy(signature[64-len(sBytes):64], sBytes)

	return signature, v, nil
}

func verify(publicKey, ID, hash, signature []byte, name string) bool {
	if publicKey == nil || len(publicKey) != 64 {
		return false
	}
	if signature == nil || len(signature) != 64 {
		return false
	}
	if hash == nil || len(hash) != 32 {
		return false
	}

	pubk := new(ecdsa.PublicKey)

	if name == "secp256k1" {
		pubk.Curve = secp256k1
	} else {
		if name == "secp256r1" {
			pubk.Curve = secp256r1
		} else if name == "sm2_std" {
			pubk.Curve = sm2_std
			pubk.X = new(big.Int).SetBytes(publicKey[:32])
			pubk.Y = new(big.Int).SetBytes(publicKey[32:])

			r := new(big.Int).SetBytes(signature[:32])
			s := new(big.Int).SetBytes(signature[32:])

			return sm2_std_verify(pubk, hash, ID, r, s)
		} else {
			return false
		}
	}
	pubk.X = new(big.Int).SetBytes(publicKey[:32])
	pubk.Y = new(big.Int).SetBytes(publicKey[32:])

	r := new(big.Int).SetBytes(signature[:32])
	s := new(big.Int).SetBytes(signature[32:])

	return verifyesdsa(pubk, hash, r, s)
}

func encrypt(pubkey []byte, plain []byte, name string) ([]byte, error) {
	if name != "sm2_std" {
		return nil, ErrUnknownCurve
	}

	if pubkey == nil || len(pubkey) != 64 {
		return nil, ErrPublicKeyIllegal
	}

	if plain == nil || len(plain) == 0 {
		return nil, ErrMessageIllegal
	}

	pubk := new(ecdsa.PublicKey)
	pubk.Curve = sm2_std
	pubk.X = new(big.Int).SetBytes(pubkey[:32])
	pubk.Y = new(big.Int).SetBytes(pubkey[32:])

	return sm2_std_encrypt(pubk, plain)
}

func decrypt(prikey, cipher []byte, name string) ([]byte, error) {

	if name != "sm2_std" {
		return nil, ErrUnknownCurve
	}

	if prikey == nil || len(prikey) != 32 {
		return nil, ErrPrivateKeyIllegal
	}

	if cipher == nil || len(cipher) == 0 {
		return nil, ErrMessageIllegal
	}

	var sm2Curve *sm2_stdCurve
	privateKeyBig := new(big.Int).SetBytes(prikey)
	priv := new(ecdsa.PrivateKey)

	priv.D = privateKeyBig
	sm2Curve = sm2_std

	if privateKeyBig.Cmp(sm2Curve.Params().N) >= 0 {
		return nil, ErrPrivateKeyIllegal
	}

	priv.PublicKey.Curve = sm2Curve
	priv.D = privateKeyBig
	priv.PublicKey.X, priv.PublicKey.Y = sm2Curve.ScalarBaseMult(prikey)

	return sm2_std_decrypt(priv, cipher)
}

func MulBaseG_Add(pointin, scalar []byte, name string) (point []byte, isinfinity bool) {

	var curve *curveParam
	var k1curve *secp256k1Curve
	var sm2Curve *sm2_stdCurve
	privateKeyBig := new(big.Int).SetBytes(scalar)
	priv := new(ecdsa.PrivateKey)
	p := new(ecdsa.PublicKey)
	p.X = new(big.Int).SetBytes(pointin[:32])
	p.Y = new(big.Int).SetBytes(pointin[32:])

	priv.D = privateKeyBig
	if name == "secp256k1" {
		k1curve = secp256k1

		priv.PublicKey.Curve = k1curve
		p.Curve = k1curve

		priv.PublicKey.X, priv.PublicKey.Y = k1curve.ScalarBaseMult(scalar)

		priv.PublicKey.X, priv.PublicKey.Y = k1curve.Add(priv.PublicKey.X, priv.PublicKey.Y, p.X, p.Y)

	} else if name == "sm2_std" {

		sm2Curve = sm2_std

		priv.PublicKey.Curve = sm2Curve
		p.Curve = sm2Curve

		priv.PublicKey.X, priv.PublicKey.Y = sm2Curve.ScalarBaseMult(scalar)

		priv.PublicKey.X, priv.PublicKey.Y = sm2Curve.Add(priv.PublicKey.X, priv.PublicKey.Y, p.X, p.Y)

	} else { // ecdsa
		if name == "secp256r1" {
			curve = secp256r1
		} else {
			return nil, false
		}

		priv.PublicKey.Curve = curve
		p.Curve = curve
		priv.D = privateKeyBig
		priv.PublicKey.X, priv.PublicKey.Y = curve.ScalarBaseMult(scalar)
		priv.PublicKey.X, priv.PublicKey.Y = curve.Add(priv.PublicKey.X, priv.PublicKey.Y, p.X, p.Y)

	}

	if priv.PublicKey.X.Cmp(big.NewInt(0)) == 0 && priv.PublicKey.Y.Cmp(big.NewInt(0)) == 0 {
		return nil, true
	}

	x := priv.PublicKey.X.Bytes()

	for len(x) < 32 {
		x = append([]byte{0x00}, x...)
	}

	y := priv.PublicKey.Y.Bytes()

	for len(y) < 32 {
		y = append([]byte{0x00}, y...)
	}
	return append(x, y...), false
}

func Add(point1, point2 []byte, name string) (point []byte, isinfinity bool) {
	var curve *curveParam
	var k1curve *secp256k1Curve
	var sm2Curve *sm2_stdCurve

	var x_big, y_big *big.Int

	p1 := new(ecdsa.PublicKey)
	p1.X = new(big.Int).SetBytes(point1[:32])
	p1.Y = new(big.Int).SetBytes(point1[32:])

	p2 := new(ecdsa.PublicKey)
	p2.X = new(big.Int).SetBytes(point2[:32])
	p2.Y = new(big.Int).SetBytes(point2[32:])

	if name == "secp256k1" {
		k1curve = secp256k1
		x_big, y_big = k1curve.Add(p1.X, p1.Y, p2.X, p2.Y)
	} else if name == "sm2_std" {
		sm2Curve = sm2_std
		x_big, y_big = sm2Curve.Add(p1.X, p1.Y, p2.X, p2.Y)
	} else if name == "secp256r1"{ // ecdsa
		curve = secp256r1
		x_big, y_big = curve.Add(p1.X, p1.Y, p2.X, p2.Y)
	} else {
		return nil, false
	}

	if x_big.Cmp(big.NewInt(0)) == 0 && x_big.Cmp(big.NewInt(0)) == 0 {
		return nil, true
	}

	x := x_big.Bytes()

	for len(x) < 32 {
		x = append([]byte{0x00}, x...)
	}

	y := y_big.Bytes()

	for len(y) < 32 {
		y = append([]byte{0x00}, y...)
	}
	return append(x, y...), false
}

func Mul(pointin, scalar []byte, name string) (point []byte, isinfinity bool) {

	var curve *curveParam
	var k1curve *secp256k1Curve
	var sm2Curve *sm2_stdCurve
	privateKeyBig := new(big.Int).SetBytes(scalar)
	priv := new(ecdsa.PrivateKey)
	p := new(ecdsa.PublicKey)
	p.X = new(big.Int).SetBytes(pointin[:32])
	p.Y = new(big.Int).SetBytes(pointin[32:])

	priv.D = privateKeyBig
	if name == "secp256k1" {
		k1curve = secp256k1
		priv.PublicKey.Curve = k1curve
		p.Curve = k1curve
		priv.PublicKey.X, priv.PublicKey.Y = k1curve.ScalarMult(p.X, p.Y, scalar)
	} else if name == "sm2_std" {
		sm2Curve = sm2_std
		priv.PublicKey.Curve = sm2Curve
		p.Curve = sm2Curve
		priv.PublicKey.X, priv.PublicKey.Y = sm2Curve.ScalarMult(p.X, p.Y, scalar)
	} else if name == "secp256r1"{ // ecdsa
		curve = secp256r1
		priv.PublicKey.Curve = curve
		p.Curve = curve
		priv.D = privateKeyBig
		priv.PublicKey.X, priv.PublicKey.Y = curve.ScalarMult(p.X, p.Y, scalar)
	} else {
		return nil, false
	}

	if priv.PublicKey.X.Cmp(big.NewInt(0)) == 0 && priv.PublicKey.Y.Cmp(big.NewInt(0)) == 0 {
		return nil, true
	}

	x := priv.PublicKey.X.Bytes()

	for len(x) < 32 {
		x = append([]byte{0x00}, x...)
	}

	y := priv.PublicKey.Y.Bytes()

	for len(y) < 32 {
		y = append([]byte{0x00}, y...)
	}
	return append(x, y...), false
}